Persistent Omnisharp in tests

mispencer opened the pull request:
Running all the tests in ycmd takes a long time. The reason for this is
repeatedly starting and stopping Omnisharp servers. To mitigated this, whenever
we start an instance of Omnisharp during testing, we cache it, and reuses it
from the cache whenever a test requires a instance of Omnisharp on the same
solution.

As part of this, add Set/GetOmnisharpPort commands to the C# completer. These
commands allow setting the port it uses to communicate with the Omnisharp
server. This allows sharing of the Omnisharp server by retrieving the port and
using it in other clients, or get the port another client is using and using it
here.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/339)
<!-- Reviewable:end -->

vheon commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9Y_2a4TibiWrLXD3vG:613514589) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
I'm not really sure about this subcommands. Why couldn't we use the already present `csharp_server_port` user option?

---

*[ycmd/completers/cs/cs_completer.py, line 387 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YXuDVwBbkNFpQeFqI:-K9YXuDVwBbkNFpQeFqJ:-592553413) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L387)):*
Is this PR a WIP or is it READY to review? if is READY remove the commented out code, if is still WIP add it to the title of the PR like [WIP] ... :+1: 

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9YYpbky3RKxiBveF6A:-672455709) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
I believe that this can be a method on the `Cs_Handlers_test` class.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 44 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YZhADyjCFto7I7ieP:-K9YZhADyjCFto7I7ieQ:1485357404) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/cs_handlers_test.py#L44)):*
Why couldn't we use the `setUp` and `tearDown` directly?

---

*[ycmd/tests/cs/cs_handlers_test.py, line 129 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YZKrW9HuDydSbh5xy:-K9YZKrXR_49MKKkMTrX:-1418941515) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/cs_handlers_test.py#L129)):*
the semicolons aren't needed :smiley: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Yay! Thanks for working on this! :D

I see this is all still a bit WIP so please change the title to have `[READY]` once you'd like a final review.

---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9Yp-yGjO_41fM6Uu4Z:132246523) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
@vheon seems to have a point here; we can't we just pass a port in options?

---

*[ycmd/tests/cs/diagnostics_test.py, line 76 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YoPHFRH5xifvHc4WU:-K9YoPHFRH5xifvHc4WV:-1939848188) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/diagnostics_test.py#L76)):*
Can we change C# test class to make this first call somehow? Having this logic repeated in tests seems icky. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9YpCwiqCRgxFX0-x6h:-562792093)*
<!-- Sent from Reviewable.io -->

mispencer commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9Yr6c3aMPttf03S-dM:1411111192) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
Because the current `csharp_server_port` can only be used for a single solution (because each solution have its own Omnisharp instance and they need to have different ports.) We need to be able to set ports for multiple solutions. 

Changing the option to support multiple solutions seems complicated, and of equally limited usage.

---

*[ycmd/completers/cs/cs_completer.py, line 387 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YXuDVwBbkNFpQeFqI:-K9Z1fBLtncImFUEoSZY:1037350481) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L387)):*
Opps, I forgot to add that back. Fixed.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9Z-aJFZWkNqoJJZesH:-1365686290) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
I don't think you can put package teardowns in the class. They have to be in the `_init_.py`

---

*[ycmd/tests/cs/cs_handlers_test.py, line 44 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YZhADyjCFto7I7ieP:-K9Yqg-ZvB7R0UYedQrX:247986582) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/cs_handlers_test.py#L44)):*
Not all tests need this - it would be wasteful to call it for every tests. Also, the multiple solution tests call it multiple times in a single test and the tests for starting/stopping the service need to be able to control the process directly.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 129 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YZKrW9HuDydSbh5xy:-K9Z1Zaj9bsFoIC5T_iL:1946294072) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/cs_handlers_test.py#L129)):*
Removed superficial semicolons

---

*[ycmd/tests/cs/diagnostics_test.py, line 76 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YoPHFRH5xifvHc4WU:-K9Z04TCsDPcMr27tibQ:-232068406) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/diagnostics_test.py#L76)):*
I don't think so. We don't have the file path to call it with in the setup. I could encapsulate it in a method, but I don't know if that would improve matters.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 6 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9ZGF0mQHHq4v6fUv2S:335621318) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
Wait, how do we handle things in "production" then? if a user has the `csharp_server_port` specified as a user option how this will be used? Is it used for the first Omnisharp Server that we start? And what would happen if he then open a different solution?

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9ZGhUxy3BHJ50Peccn:-72387644) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
http://nose.readthedocs.org/en/latest/writing_tests.html#test-classes

looks like if the method have the right name is used right. Am I reading the doc in the wrong way?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@puremourning You have excellent instincts & experience when it comes to test-writing; would you mind giving this a review? Thanks! :)

---

Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9ZHqM4AY0RTmk9uWsw:-824818157) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
Just making a note that there's commented-out code here so you don't forget to sort it out. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9ZHfTY-yNXVPcYJmM6:-1122720510)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I will yes of course. However I can't promise it won't be until the weekend. My work life is 50/50 code reviews and design meetings at the moment so I'm a bit frazzled :) this one looks like one that needs a bit of focus. 

At a high level I wonder if we shouldn't be using a real ycmd instance for this sort of thing rather than changing the code to help test scenarios. 

But then again, I'm sick of waiting for csharp tests to run. And the full integration test sounds hard to manage. Sorry, waffling. 

---

Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9ZIBDjQPixF6V5vXeF:653360205)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@puremourning Oh man, only meetings and code reviews? :D That must suck. You know what I say about OSS dev work: take it slow, don't burn out. When you have the time to look at it, please do. 

If you end up not being able/willing to review it, tell us and we'll manage. :)

---

Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9ZMGboe2X3UITm63SO:1848468579)*
<!-- Sent from Reviewable.io -->

puremourning commented:
crunching x while planning to crunch x+1 

---

Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9ZOszZaiqzZGF3XiNL:-412892788)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Par for the course really :D

---

Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9ZPCF_H91H0IaiUtID:214209316)*
<!-- Sent from Reviewable.io -->

mispencer commented:
Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9ZYSoKyxIz7rQpRHXF:-358321738) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
If you set the `csharp_server_port` option, you can't use the multiple solution support. Fixed port is just not compatible with multiple Omnisharp instances.  So, yes, you can only use ycmd on the first solution. I am not sure what happens if you try to open another solution, but probably nothing good.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9ZYlE62Hmz33r3baJG:983031507) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
Well, I tested `teardown_package` in the test class, and it wasn't called. It was only called if defined in the `_init_.py` file. We can define  a class teardown (`teardown_all`), but then we wouldn't be sharing Omnisharp instances between the three test classes.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9ZZJPf3lQpZdePwpLQ:-1098810297) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
Yes, my original plan when I implemented this a few months ago was to preserve the existing behavior in CI. Should I just remove this? I was wondering if it should possible to run tests without the Omnisharp instance caching (and other semantic servers if implemented) in case of flakiness around it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[ycmd/completers/cs/cs_completer.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9Y_2a3zCFQiu93AQv2:-K9cxytkQqnCJ0y4vDlP:-1184468408) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/completers/cs/cs_completer.py#L190)):*
it will try to use the same port an it would fail to start the second instance... this is bad :( But since we didn't saw any bug report users either don't use those options or they don't use the multiple solution support...

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9cxOVvHjnHnNO6kjwx:-1313534250) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
IMHO it would be ok to share the Omnisharp instances only between test of the same test class but you've already did this, so is ok :+1: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9dCzYPyGSHmNI7kj5B:-381681171) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
I think we always want to use one instance of omnisharp. If a specific test needs its own instance, we make that one test do that.

If we have flakiness with omnisharp, we need to resolve it no matter what.

If we want to temporarily disable C# tests, we can do that with flags to run_tests.py but I'd rather avoid it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9e_anUaAVNnbC7tIL7:380505225) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
@mispencer Let's try having only class-level omnisharp instances. The code will be less surprising that way (few think to look into `__init__.py` files) and we should still get 95% of the speed benefit. We have a ton of C# tests but only 3 test classes. Should be good enough.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

mispencer commented:
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9hDzXgHsRG2MvE-Xz1:-1869884358) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
Class-level Omnisharp instances increases the test run time by 20 seconds on my machine. This isn't a issue for CI, but it's enough to be irritating when running tests manually. So I just added a comment in the cs_handler_test class to look here for teardown_package. 

---

*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9hGCy_k1HvuqNA-l0t:-261606086) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
I added a toggle for this. I don't expect CI will used this, but I think it may be useful locally for debugging. However, I'm not attached to this - let me know what you think.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9YYpbky3RKxiBveF69:-K9mRKI8Ja5eD-ZxBXNq:1140080906) ([raw file](https://github.com/valloric/ycmd/blob/80220d7b4c752766f5631cd76c35dfd0f3b3afee/ycmd/tests/cs/__init__.py#L3)):*
OK, sold. :)

---

*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9mRTBFnSq_igGhkdGQ:1978316548) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
I'm fine with this, but do add a comment on `INSTANCE_PER_TEST` that explains it a bit.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@mispencer You frequently send PR's to YCM & ycmd (which we love!); would you like to be added to the invite-only ycm-dev mailing list? If yes, shoot me an email (my email address is on [my about page](https://val.markovic.io/about)).

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9njiMD6-Sst0k9nXWM:-1112392243)*
<!-- Sent from Reviewable.io -->

mispencer commented:
*[ycmd/tests/cs/cs_handlers_test.py, line 65 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9ZHqM3wuu7RgUtRsp8:-K9nuHYfHZbJUjZYbI8N:1139853271) ([raw file](https://github.com/valloric/ycmd/blob/855f8e3d2a5f1a8f0391bdec550245721a032e17/ycmd/tests/cs/cs_handlers_test.py#L65)):*
`INSTANCE_PER_TEST` is now documented.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
A couple of minor comments from, but other than that, I'm <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> on this. We'll wait for @puremourning's review unless he tells us otherwise.

I'm super excited about this change; our Travis runs should go sooo much faster! :D

---

Reviewed 4 of 6 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 558 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9oFmGmipVTZEns2e7S:-K9oFmGmipVTZEns2e7T:-1870709436) ([raw file](https://github.com/valloric/ycmd/blob/37cfcf9fbe6e21f454929a8aecf144389bc1a9b1/ycmd/completers/cs/cs_completer.py#L558)):*
Instead of a raw False, maybe call it with `external_check = False`? This way it's clear what's going on.

---

*[ycmd/completers/cs/cs_completer.py, line 569 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9oFtw90ybWqcWjv2mZ:-K9oFtwA8xkZtpyQQ4aJ:1121484540) ([raw file](https://github.com/valloric/ycmd/blob/37cfcf9fbe6e21f454929a8aecf144389bc1a9b1/ycmd/completers/cs/cs_completer.py#L569)):*
Similar to above.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9oGPHgzS239vNwAuZW:-237798273)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Its in my fun things to do on a Saturday list :)

> On 6 Feb 2016, at 01:52, Val Markovic <notifications@github.com> wrote:
> 
> A couple of minor comments from, but other than that, I'm  on this. We'll wait for @puremourning's review unless he tells us otherwise.
> 
> I'm super excited about this change; our Travis runs should go sooo much faster! :D
> 
> Reviewed 4 of 6 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
> Review status: all files reviewed at latest revision, 2 unresolved discussions.
> 
> ycmd/completers/cs/cs_completer.py, line 558 [r4] (raw file):
> Instead of a raw False, maybe call it with external_check = False? This way it's clear what's going on.
> 
> ycmd/completers/cs/cs_completer.py, line 569 [r4] (raw file):
> Similar to above.
> 
> Comments from the review on Reviewable.io
> 
> —
> Reply to this email directly or view it on GitHub.
>

mispencer commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 558 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9oFmGmipVTZEns2e7S:-K9oN7Fee4ds3OHf8Mx7:886438527) ([raw file](https://github.com/valloric/ycmd/blob/37cfcf9fbe6e21f454929a8aecf144389bc1a9b1/ycmd/completers/cs/cs_completer.py#L558)):*
Explicitly labeled argument for ServerIsRunning

---

*[ycmd/completers/cs/cs_completer.py, line 569 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9oFtw90ybWqcWjv2mZ:-K9oN9fS4AEKM83RcFHY:-1213865979) ([raw file](https://github.com/valloric/ycmd/blob/37cfcf9fbe6e21f454929a8aecf144389bc1a9b1/ycmd/completers/cs/cs_completer.py#L569)):*
Here too.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Couple of minor comments, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[ycmd/completers/cs/cs_completer.py, line 193 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9qoT20CigND3Woka-Y:-K9qoT20CigND3Woka-Z:-95491041) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/completers/cs/cs_completer.py#L193)):*
Not a big deal, but this will throw a generic range error if the user doesn't supply a port, rather than a human-readable one.

---

*[ycmd/completers/cs/cs_completer.py, line 594 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9qrKhT0JwmfkOkzdTk:-K9qrKhT0JwmfkOkzdTl:375945184) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/completers/cs/cs_completer.py#L594)):*
shouldn't this method be (strictly) protected by the lock?

---

*[ycmd/tests/cs/diagnostics_test.py, line 33 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9qu2RZ18v7VwsoFiNe:-K9qu2RZ18v7VwsoFiNf:51755368) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/diagnostics_test.py#L33)):*
Is there any more explanation for this? I find it irksome that we don't know why this is happening and just working around it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9qtxAcw6OZlhlyz9VZ:-1640992660)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 4 of 6 files at r1, 1 of 1 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9rYsYAGNt21_0HxXha:-957377383)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 4 of 6 files at r1, 1 of 1 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 198 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rwsBFOabpjIRcpuNR:-K9rwsBGflsnLMCKC6Wu:-918530849) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/completers/cs/cs_completer.py#L198)):*
`: ( lambda self, request_data, args:` should be aligned with above ones.

---

*[ycmd/completers/cs/cs_completer.py, line 550 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rtEbtpCjF2YcinEgQ:-K9rtEbtpCjF2YcinEgR:-2062185516) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/completers/cs/cs_completer.py#L550)):*
This `else:` is not needed.

---

*[ycmd/completers/cs/cs_completer.py, line 552 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rtIpJOZ37YdrdEvbS:-K9rtIpJOZ37YdrdEvbT:-918288111) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/completers/cs/cs_completer.py#L552)):*
This one too.

---

*[ycmd/tests/cs/__init__.py, line 1 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pvVZ6GhARHRgMhpN1:-K9pvVZ6GhARHRgMhpN2:-1291335127) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/__init__.py#L1)):*
Missing copyright header.

---

*[ycmd/tests/cs/__init__.py, line 2 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rxq1c0pPRuWiRr7W2:-K9rxq1c0pPRuWiRr7W3:-657335852) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/tests/cs/__init__.py#L2)):*
Two blank lines.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rung7O0GZPpi3Bg5P:-K9rung9tdvcMYCnxFx4:687615381) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/tests/cs/__init__.py#L3)):*
Can be renamed `tearDownPackage` to be closer to functions naming style.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pvgZ-E21TDQvAgwUp:-K9pvgZ-E21TDQvAgwUq:749008343) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L49)):*
`See __init__.py file` would be more clear.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 94 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9q-I7bpGvbxeF1E2bf:-K9q-I7bpGvbxeF1E2bg:256394868) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L94)):*
We can return here and remove the `else:`.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 124 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pzsCzXFwWcgZURmyR:-K9pzsD-1uf6tKrNZeEy:-1866922122) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L124)):*
No space before `(`. Could be directly returned.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 170 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9q-vs9SiJHyNZrLuv3:-K9q-vsA41_t9ducDIb0:-1244129947) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L170)):*
This could be a `@classmethod` in `Cs_Handlers_test`.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 31 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rxnkegrGeBemOYqdw:-K9rxnkfdgSkyYKXtv5W:-657335852) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/tests/cs/cs_handlers_test.py#L31)):*
Two blank lines.

---

*[ycmd/tests/cs/diagnostics_test.py, line 37 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pxNMPndK5BceZEfFT:-K9pxNMQO-6BRpS6dOMP:-1823556964) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/diagnostics_test.py#L37)):*
Parameters alignment.

---

*[ycmd/tests/cs/diagnostics_test.py, line 80 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pxR_0KfX6dKt6jhpY:-K9pxR_0KfX6dKt6jhpZ:-1823556964) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/diagnostics_test.py#L80)):*
Parameters alignment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

mispencer commented:
Fixed whitespace and alignment.

---

Review status: 2 of 6 files reviewed at latest revision, 15 unresolved discussions.

---

*[ycmd/completers/cs/cs_completer.py, line 594 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9qrKhT0JwmfkOkzdTk:-K9taVc_3f0UZTBhtW92:-36493614) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/completers/cs/cs_completer.py#L594)):*
`_SetOmnisharpPort` now is wrapped by the  server state lock.

---

*[ycmd/completers/cs/cs_completer.py, line 550 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rtEbtpCjF2YcinEgQ:-K9t_9iOfazarIPZHu6h:-1557053133) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/completers/cs/cs_completer.py#L550)):*
`ServerIsRunning` has been refactored to remove elses.

---

*[ycmd/tests/cs/__init__.py, line 1 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pvVZ6GhARHRgMhpN1:-K9t_IjVBVnTfAZr2cEm:-1755670013) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/__init__.py#L1)):*
Added copyright header.

---

*[ycmd/tests/cs/__init__.py, line 3 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9rung7O0GZPpi3Bg5P:-K9t_LgFdbuiYo5Xl68B:-1882438762) ([raw file](https://github.com/valloric/ycmd/blob/e6cbf3db86ecc406f0e0f5745d490e7d9e0c3e07/ycmd/tests/cs/__init__.py#L3)):*
Renamed to `teardownPackage`.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pvgZ-E21TDQvAgwUp:-K9t_PqZt4SKguMfU5wA:1335273161) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L49)):*
Spelled out `__init__.py`

---

*[ycmd/tests/cs/cs_handlers_test.py, line 94 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9q-I7bpGvbxeF1E2bf:-K9t_0SgJNU-5nMrsxqb:-1086735154) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L94)):*
`_FindOmniSharpSolutionPath` has been refactored to remove elses.

---

*[ycmd/tests/cs/cs_handlers_test.py, line 124 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9pzsCzXFwWcgZURmyR:-K9t_x2RkqWN-hYSjBDF:-2976918) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/cs_handlers_test.py#L124)):*
`_FindOmniSharpSolutionPath` has been refactored to return directly.

---

*[ycmd/tests/cs/diagnostics_test.py, line 33 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/339#-K9qu2RZ18v7VwsoFiNe:-K9tZjmcbx4rrdMe6NDB:-1776471401) ([raw file](https://github.com/valloric/ycmd/blob/550f637851dcf3984c63a679c972c0e73aa2c396/ycmd/tests/cs/diagnostics_test.py#L33)):*
 I don't have any explanation. My attempt to debug was fruitless. In any case, it's not introduced by this PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9t_DRLJ8Ra6xgGDbuf:719055142)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

micbou commented:
Squash commits and <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9uvUbgH4JSH-p6Ajpq:1105808245)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339)*
<!-- Sent from Reviewable.io -->

mispencer commented:
It's been squashed to two commits and rebased on master.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/339#-:-K9x1uSbb4SKOlerNLCq:1108965663)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Thanks for the PR! :)

@homu r+

homu commented:
:pushpin: Commit 4638bac has been approved by `Valloric`

<!-- @homu r=Valloric 4638bace3580c49d9e4e0bd8b1858416717fabf1 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/107624189)
